src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1707,32 +1718,59 @@ internal void PerformUserCancellation()
             if (!ReadingPrependedMessagesMRE.Wait(0))
                 return;
 
-            _userCancellationRequested = true;
-
-            if (AttemptPostgresCancellation && SupportsPostgresCancellation)
-            {
-                var cancellationTimeout = Settings.CancellationTimeout;
-                if (PerformPostgresCancellation() && cancellationTimeout >= 0)
-                {
-                    if (cancellationTimeout > 0)
-                    {
-                        ReadBuffer.Timeout = TimeSpan.FromMilliseconds(cancellationTimeout);
-                        ReadBuffer.Cts.CancelAfter(cancellationTimeout);
-                    }
+            PerformUserCancellation();
+        }
+        finally
+        {
+            Monitor.Exit(CancelLock);
+        }
+    }
 
-                    return;
-                }
-            }
+    void PerformDelayedUserCancellation()
+    {
+        // Take the lock first to make sure there is no concurrent Break.
+        // We should be safe to take it as Break only take it to set the state.
+        lock (SyncObj)
+        {
+            // The connector is dead, exit gracefully.
+            if (!IsConnected)
+                return;
+            // The connector is still alive, take the CancelLock before exiting SingleUseLock.
+            // If a break will happen after, it's going to wait for the cancellation to complete.
+            Monitor.Enter(CancelLock);
```

Comment:
Any way this can deadlock on another cancellation?

Reply:
Don't think so? There are only 3 places where we take `CancelLock` without checking whether we have `SyncObj` lock:

1. `ResetCancellation`, which is called every time we send a query and it just resets `ReadingPrependedMessagesMRE`
2. `PerformPostgresCancellation`, but that's the method doing the actual cancellation work (that is, it creates a new connector and sends a cancellation request), so it should block until it's either done or hits a timeout.
3. `Break`, but it takes `SyncObj` first and then changes the state to `Broken`, so cancellation should immediately exit after it takes `SyncObj` lock and checks `IsConnected`.